src/test/java/com/google/sps/MapServletTest.java

```diff
@@ -0,0 +1,186 @@
+
+  @Before
+  public void initMapServlet() {
+    mapServlet = new MapServlet();
+  }
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  /* 
+   * Tests that doGet constructs the expected JSON string containing locations in order
+   */
+  @Test
+  public void testWriteLocations() throws Exception {
+    HttpServletRequest request = mock(HttpServletRequest.class);       
+    HttpServletResponse response = mock(HttpServletResponse.class);    
+
+    // Create writers to check against actual output.
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    // initialize datastore
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // create user entity
+    Entity userEntity = new Entity(User.USER);
+    userEntity.setProperty(User.USER_EMAIL, EMAIL);
+    datastore.put(userEntity);
+    Key userEntityKey = userEntity.getKey();
+
+    // Add a single Trip to Datastore with the User Entity Key.
+    Entity tripEntity = new Entity(Trip.TRIP, userEntityKey);
+    tripEntity.setProperty(Trip.TRIP_NAME, TRIP_NAME);
+    tripEntity.setProperty(Trip.DESTINATION_NAME, INPUT_DESTINATION);
+    tripEntity.setProperty(Trip.IMAGE_SRC, IMAGE_SRC);
+    tripEntity.setProperty(Trip.START_DATE, TRIP_DAY_OF_TRAVEL);
+    tripEntity.setProperty(Trip.END_DATE, TRIP_DAY_OF_TRAVEL);
+    datastore.put(tripEntity);
+
+    // create tripDay entity
+    Entity tripDayEntity = new Entity(TripDay.QUERY_STRING, tripEntity.getKey());
+    tripDayEntity.setProperty("origin", INPUT_DESTINATION);
+    tripDayEntity.setProperty("destination", INPUT_DESTINATION);
+    tripDayEntity.setProperty("date", TRIP_DAY_OF_TRAVEL);
+    datastore.put(tripDayEntity);
+
+    // create location entities and put it into datastore
+    Entity DomeEntity = new Entity(TripDay.LOCATION_ENTITY_TYPE, tripDayEntity.getKey());
+    DomeEntity.setProperty(TripDay.NAME, DOME_ADDRESS);
+    DomeEntity.setProperty(TripDay.ORDER, 1);
+    datastore.put(DomeEntity);
+
+    Entity YosemiteEntity = new Entity(TripDay.LOCATION_ENTITY_TYPE, tripDayEntity.getKey());
+    YosemiteEntity.setProperty(TripDay.NAME, YOSEMITE_ADDRESS);
+    YosemiteEntity.setProperty(TripDay.ORDER, 0);
+    datastore.put(YosemiteEntity);
+
+    // run do Get
+    mapServlet.doGetMap(response, datastore, userEntity, tripEntity.getKey());
+
+    // even though DomeEntity is added first, its order property is 1 so it should appear 
+    // after YosemiteEntity in the JSON.
+    String expectedJson = "[\"" + INPUT_DESTINATION + "\",\""+YOSEMITE_ADDRESS+"\",\""+DOME_ADDRESS+"\"]";
+    
+    writer.flush();
+    Assert.assertTrue(stringWriter.toString().contains(expectedJson));
+  }
+
+  /* 
+   * Tests doGet when there is no current user
+   */
+  @Test
+  public void testNoCurrentUser() throws Exception {
```

Comment:
Added! I mocked the AuthServlet class with PowerMockito so that the logged in user is user2. I then added a trip under user1 in datastore and mocked the request parameter to be that tripKey. This returned a "no trip found" message as expected.